### DISCUSSION TITLE
FR: Site specific email settings

### DISCUSSION THREAD

**Discussion by Mosnar** [craftcms/cms#8709] [Ideas] [answered]
### Description

Title says it all. With support for multiple domains, it doesn't make a lot of sense to have emails come from the same sender for certain types of emails. For example, if I do a password reset from siteb.com, I would get an email from the sitea.com system email address, which is quite confusing/alarming to siteb.com users.

### Additional info

- Craft version: 3.1.34
- PHP version: N/A
- Database driver & version: N/A
- Plugins & versions: N/A

**Comment by Mosnar** [comment]
For others trying to do this, it's actually really easy to accomplish; however, I still think it's worth leaving this issue open, as it's not very intuitive for those without a lot of experience with Craft.

In our `config/app.php` we just need to override the mailer component of Craft. So in the 'components' key of the app.php file (create one if it doesn't exist), you just need something like this (for Mailgun):

```php
'mailer' => function() {
            // Get the stored email settings
            $settings = craft\helpers\App::mailSettings();

            // Override the transport adapter class
            $settings->transportType = craft\mailgun\MailgunAdapter::class;

            // Override the transport adapter settings
            if (Craft::$app->sites->getCurrentSite()->handle === 'site1') {
                $settings->transportSettings = [
                    'domain' => getenv('MG_DOMAIN_SITE1'),
                    'apiKey' => getenv('MG_API_KEY_SITE1'),
                ];
                $settings->fromEmail = getenv('SITE1_FROM_EMAIL');
                $settings->fromName = getenv('SITE1_FROM_NAME');
            } else {
                $settings->transportSettings = [
                    'domain' => getenv('MG_DOMAIN_SITE2'),
                    'apiKey' => getenv('MG_API_KEY_SITE2'),
                ];
                $settings->fromEmail = getenv('SITE2_FROM_EMAIL');
                $settings->fromName = getenv('SITE2_FROM_NAME');
            }

            // Create a Mailer component config with these settings
            $config = craft\helpers\App::mailerConfig($settings);

            // Instantiate and return it
            return Craft::createObject($config);
        },
```

**Reply by thomasgoemaere** [reply]
Hi @Mosnar i've used your code in my Craft3 multisite but it seems to ignore the fromEmail & fromName setting (transportSettings are correctly overwritten). For the from email & name it seems to always use the ones defined in de .env file (FROM_NAME & FROM_EMAIL) and it doesn't use the values I provide in the mailer function. Does the code work correctly in your site? Any ideas of what i'm doing wrong? Thank you.

**Comment by putyourlightson** [comment]
The UI for allowing a full range of site specific email sending settings in the control panel might be tricky. What might be more feasible is to add site specific email addresses, while maintaining a single set of sending settings. I've done something similar in the Campaign plugin and have mocked up what it might look like in Craft.

<img width="840" alt="Screenshot 2019-07-17 at 20 06 00" src="https://user-images.githubusercontent.com/2318222/61399539-a2d35900-a8ce-11e9-9b39-95af04bdaf64.png">

If this is a feature that the P&T team would like to add then I'd be happy to submit a pull request.

**Comment by brandonkelly** [comment]
Yeah maybe. I worry that it’s going to be annoying for sites where no site-specific values are needed though.

**Comment by theskyfloor** [comment]
Just fall back to the default if the settings are blank?

**Comment by putyourlightson** [comment]
@theskyfloor Where is the default set?

**Comment by theskyfloor** [comment]
"Site A" in which the handle is usually automatically set to default. Much like this:

"Leave blank if settings should match default site"

<img width="750" alt="Screen Shot 2019-07-18 at 12 11 43 PM" src="https://user-images.githubusercontent.com/8887839/61477605-48311000-a955-11e9-9bf0-53fc648d4553.png">

**Comment by putyourlightson** [comment]
So are you implying that the email settings would be required for the primary site (the first row in the table)?

**Comment by theskyfloor** [comment]
Default site settings (first row) would follow the same exact convention as now. Email would be filled in with the administrator email, from name with the site name, template blank:

<img width="519" alt="Screen Shot 2019-07-18 at 12 44 02 PM" src="https://user-images.githubusercontent.com/8887839/61479608-cc859200-a959-11e9-83af-f7a40a6822f2.png">

**Comment by putyourlightson** [comment]
My proposal was to actually replace the first 3 fields in your screenshot with the site specific email settings table. I'm not sure it makes sense to have them both. 

<img width="840" alt="Screenshot 2019-07-17 at 20 06 00" src="https://user-images.githubusercontent.com/2318222/61479838-fb6d2980-a994-11e9-8cde-dc0bbd93790e.png">

**Comment by theskyfloor** [comment]
@putyourlightson totally! I was just posting that as an example of what Craft already does by default! I think your proposal is brilliant and spot on... the first row would just act as the current 3 fields do just in a different format... new sites default to the Primary Site unless otherwise defined :)

**Comment by putyourlightson** [comment]
Got it, so we _are_ on the same page!!

**Comment by brandonkelly** [comment]
That wouldn’t work for optional settings, like Email Template, where it wouldn’t be clear whether a blank value meant “no email template for this site” or “use the primary site’s template”.

**Comment by putyourlightson** [comment]
That's true. So then let's scrap the idea of falling back to the primary site's values for empty fields and make the required fields required for each site. Perhaps when a new site is created it could automatically inherit the primary site's values to avoid having to manually re-enter them.

An alternative UI could be to move the 3 email setting fields into the site edit page. The question then becomes whether it is intuitive to find email related settings in the site settings.

<img width="553" alt="Screenshot 2019-07-19 at 17 38 41" src="https://user-images.githubusercontent.com/2318222/61547651-177fd200-aa4c-11e9-9aae-1d02becd7555.png">

**Comment by theskyfloor** [comment]
Same thoughts from the peanut gallery! No matter what supporting multiple domains for email seems like an essential part of offering multisite. Site settings is a pretty logical place for site specific config stuff IMO.

**Comment by hhhc** [comment]
I really would like to see this implemented as this would just complete the multi-site offering of CraftCMS.
This should then also work with SES plugin.

**Comment by AndrewMac** [comment]
I would just like to add the biggest vote of support I possibly can for getting this sorted out, please. It's a source of confusion and sometimes mistrust for site users, who are these days getting more and more careful about emails they don't recognise.

I can see that it seems to be getting pushed back from milestone to milestone, and I can see that in those milestones it competes with a lot of other great features that we're all very much looking forward to, but this is a basic service to end users, and they're what our sites (and our jobs) are all about :-)

Cheers, Andrew

**Comment by AndrewMac** [comment]
Yes, me again, because this is still a big issue for my client ...

Although this has all gone a bit quiet, I still think it's a basic need: a client of mine has related businesses that use different domains, and they really don't think it's a good look when people receive an email but don't recognise where it came from. My client really doesn't like it, and I agree with them.

We really should be able to specify that system emails come from a site, rather than from the system, which means specific email addresses for each site, and the option of different templates for each site (for branding, etc). Doesn't that make sense?

I recommend Craft to my clients because of the usually superb localisation features, so this issue just doesn't sit well. Along with the lack of site-specific Commerce stores, of course, which has lost me several jobs and is also taking a long time to be addressed. 

I haven't given up on either of these enhancements, though, so can anyone confirm whether this will ever be worked on? Thanks!

**Reply by brandonkelly** [reply]
@AndrewMac Definitely want to get to it, but we couldn’t make it happen in time for Craft 4. I’ll throw it on the list for Craft 5.

**Reply by AndrewMac** [reply]
Thanks @brandonkelly, 

Glad to hear it's not been forgotten about.
If Craft 5 delivers this, and Commerce 5 comes with multistore/multicurrency, then you will have made a lot of us happy :-)

Cheers, Andrew

**Reply by AndrewMac** [reply]
> @AndrewMac Definitely want to get to it, but we couldn’t make it happen in time for Craft 4. I’ll throw it on the list for Craft 5.

Hi Brandon, sorry to chase you on this, because I know you've got so many other great plans for Craft, but … any updates you can share? It's been on the Roadmap for so long now, and it may not be as exciting as some of the other planned features, but it's just so important to our clients and to their relationships with their user base. So, any chance of it coming soon, pretty please? ;-)

**Reply by brandonkelly** [reply]
@AndrewMac Sorry it took a lot longer than expected, but [finally got around to this](https://github.com/craftcms/cms/discussions/8709#discussioncomment-11345916) for 5.6 🎉

**Reply by AndrewMac** [reply]
Thanks so much, @brandonkelly. That's really great news! :-)

**Comment by BenParizek** [comment]
As it isn't mentioned above, another use case I would find useful here is the ability to set the Transport Type via Environment Variable (i.e. `CRAFT_TRANSPORT_TYPE`)

It seems most other settings on the Email Settings page allow this and would allow for better customization in development workflows where you may want to override the default Email Settings to use, say, DDEV's Mailhog integration (which requires SMTP, so if you are using anything other than SMTP having local settings to use Mailhog requires additional steps)

**Reply by leevigraham** [reply]
I do this in a testing environment:

```php
# app.config

<?php

use craft\helpers\App;

$config = [
    'id' => App::env('CRAFT_APP_ID') ?? 'CraftCMS',
    'modules' => [
        'site' => \modules\site\Module::class,
    ],
    'bootstrap' => [
        'site',
    ],
    'components' => [
        'queue' => [
            'ttr' => 60*10,
        ],
    ],
];

if(App::env('CRAFT_ENVIRONMENT') === 'test') {
    $config['components']['mailer'] = static function () {
        $config = craft\helpers\App::mailerConfig();
        $config['useFileTransport'] = true;
        return Craft::createObject($config);
    };
}

return $config;

```

**Comment by brandonkelly** [comment, marked as answer]
Added this for Craft 5.6 :tada: (#16187).

Also added the ability to track which site users should be primarily affiliated with, so site-specific email settings will be factored in when sending emails from the control panel as well. (#16174)

**Reply by brandonkelly** [reply]
Craft 5.6.0 is out with those changes 🎉

**Comment by satcreative** [comment]
Is there a solution for multisite Transport type and their related settings. 

I'm using the first party Mailgun plugin and have attempted @Mosnar (6 year old) suggestion of editing the 'mailer' in config/app.php, however, I get:

`Setting read-only property: craft\web\Application::mailer`

Mailgun requires that you create a domain and then create an API sending key associated to that domain, so using the same 'Domain', 'API key' and potentially 'Endpoint', would mean that at least one site is sending emails via the wrong domain. That may then result in authentication issues as the sending email address wouldn't match the Mailgun domain.

@brandonkelly It would be good to achieve this in the CP settings in the long term, but do you have a config work around that would work now?

Thanks!

**Reply by brandonkelly** [reply]
That code example should still work, but it needs to be placed within the `components` key:

```php
return [
    // ...
    'components' => [
        'mailer' => // ...
    ],
];
```

We’ll go through our first party mailer plugins and see if we can add native multi-site support to their settings.

**Reply by satcreative** [reply]
Thanks. That's working now.
And that's good to hear for the future. Cheers!
